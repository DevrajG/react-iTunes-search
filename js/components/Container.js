// @flow
import React, { Component } from 'react';
// $FlowFixMe
import reqwest from 'reqwest';
import List from './List';
import Message from './Message';
import emitter from '../emitter';
import { getMedia } from '../utils';

type ContainerState = {
  type: 'start' | 'loading' | 'noContent' | 'error',
  response?: Object
};

class Container extends Component {

  state: ContainerState = { type: 'start' };

  componentDidMount() {
    emitter.on('search', ({ media, query }) => {
      this.setState({ type: 'loading' });
      reqwest({
        url: `https://itunes.apple.com/search?media=${getMedia(media || 'all')}&term=${query.split(' ').join('+')}`,
        type: 'jsonp'
      })
      .then(response => this.setState({ response, type: response.resultCount || 'noContent' }))
      .fail(err => this.setState({ type: 'error' }));
    });
  }

  componentWillUnmount() { // eslint-disable-line class-methods-use-this
    emitter.removeListener('search');
  }

  mainRender = () => {
    const { type, response } = this.state;
    const msgMap = {
      start: {
        headerMsg: 'Welcome back!',
        iconColor: 'black',
        icon: 'help',
        bodyMsg: 'Please use enter to start search!'
      },
      loading: {
        headerMsg: 'Just one second',
        iconColor: 'blue',
        icon: 'notched circle loading',
        bodyMsg: 'Fetching data......'
      },
      noContent: {
        headerMsg: 'No search results',
        iconColor: 'yellow',
        icon: 'warning',
        bodyMsg: 'There is no data.'
      },
      error: {
        headerMsg: 'Error',
        iconColor: 'red',
        icon: 'warning sign',
        bodyMsg: 'We\'re sorry please try again later.'
      }
    };

    const msg = msgMap[type];

    if (msg) {
      return (<Message {...msg} />);
    }
    return (<List {...response} />);
  };

  render() {
    return (
      <div
        style={{
          margin: '50px auto',
          maxWidth: '900px'
        }}
      >
        {this.mainRender()}
      </div>
    );
  }

}

export default Container;
