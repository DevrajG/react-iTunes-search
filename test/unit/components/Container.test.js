import React from 'react';
import { shallow } from 'enzyme';
import Container from '../../../src/components/Container';

describe('<Container />', () => {
  let component;
  beforeEach(() => {
    component = shallow(<Container />);
  });
  test('init state', () => {
    expect(component).toMatchSnapshot();
  });
  test('loading state', () => {
    component.setState({ status: 'loading' });
    expect(component).toMatchSnapshot();
  });
  test('show data', () => {
    const data = {
      resultCount: 2,
      results: [
        {
          trackName: 'test1'
        },
        {
          trackName: 'test2'
        }
      ]
    };
    component.setState({
      status: '',
      data
    });
    expect(component).toMatchSnapshot();
  });
  test('no search results state', () => {
    component.setState({ status: 'noContent' });
    expect(component).toMatchSnapshot();
  });
  test('error state', () => {
    component.setState({ status: 'error' });
    expect(component).toMatchSnapshot();
  });
});
