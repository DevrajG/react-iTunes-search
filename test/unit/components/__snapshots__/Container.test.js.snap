// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container /> error state 1`] = `
<div>
  <Message
    status="error"
  />
</div>
`;

exports[`<Container /> init state 1`] = `
<div>
  <Message
    status="init"
  />
</div>
`;

exports[`<Container /> loading state 1`] = `
<div>
  <Message
    status="loading"
  />
</div>
`;

exports[`<Container /> no search results state 1`] = `
<div>
  <Message
    status="noContent"
  />
</div>
`;

exports[`<Container /> show data 1`] = `
<div>
  <List
    resultCount={2}
    results={
      Array [
        Object {
          "trackName": "test1",
        },
        Object {
          "trackName": "test2",
        },
      ]
    }
  />
</div>
`;
